### PR TITLE
SQL: Test and fix the NULL handling of the String functions

### LINKED ISSUE #58907: SQL: fix string functions handling of NULL parameter values
Some string functions that take an integer parameter - like SUBSTRING or INSERT - exhibit a potentially unexpected behavior when any of the integer parameters is computed from a column containing `NULL` values: the function simply returns the original string [expression] value, applying no operation.
Example: `SELECT SUBSTRING(str_column, int_column_with_nulls, some_lenght) = str_column` will return `true` every`WHERE int_column_with_nulls IS NULL`.

This is different from directly providing a `NULL` literal for any of the parameters, when the function returns `NULL` (as expected).

Note that other RDBMSes (ex: SQL Server, Postgres, MariaDB) return `NULL` when any parameter - either literal or attribute - is `NULL`; we should arguably follow lead or at least document the behavior.

### PR DESCRIPTION
**Pull request by palesz**
Fixed the inconsistencies regarding NULL argument handling.
NULL literal vs NULL field value as function arguments in some case
resulted in different function return values.

Functions should return with the same value no matter if the argument(s)
came from a field or from a literal.

The introduced integration test tests if function calls with same
argument values (regardless of literal/field) will return with the
same output (also checks if newly added functions are checked
by the test).

Fixed the following functions:
* Insert: NULL start, length and replacement arguments (as fields) also
result in NULL return value instead of returning the input.
* Locate: NULL pattern results in NULL return value
* Replace: NULL pattern and replacement results in NULL instead of
returning the input
* Substring: NULL start or length results in NULL instead of returning
the input

Fixes #58907

### CODE REVIEW

File: x-pack/plugin/sql/qa/server/src/main/resources/functions.csv-spec [hunk 1]
```diff
@@ -347,8 +347,8 @@ SELECT CAST(LOCATE(CASE WHEN languages > 3 THEN 'a' ELSE 'b' END, CASE WHEN TRUN
 
       x:s      |       c:l       
 ---------------+---------------
-false          |80             
-true           |16             
+false          |83             
+true           |13
```
Comment (palesz) [inline x-pack/plugin/sql/qa/server/src/main/resources/functions.csv-spec:351]: Especially curious about your opinions on this one:

These changes are the result of `start=NULL{f}` argument handling (this is an explicit `start=NULL` vs leaving out the optional parameter).

Previously:
`LOCATE(pattern,input,start=NULL{field})` was handled as `LOCATE(pattern,input,start=1)` or `LOCATE(pattern, input)`, but `LOCATE(pattern,input,start=NULL{literal})` resulted in `NULL`.

Now in both cases we return with 0 to be in-sync with the MySQL behaviour (where `LOCATE` comes from).

MySQL behaviour: `LOCATE(pattern,input,start=NULL) => 0`
MS SQL behaviour: `CHARINDEX(pattern,input,start=NULL) => NULL`

Comment (matriv) [reply x-pack/plugin/sql/qa/server/src/main/resources/functions.csv-spec:351]: Overall I think it would follow the MySQL behaviour since `LOCATE` comes from there.
If we had to change the behaviour only for a literal NULL I would say to go ahead and provide it as a bug fix, but since now we changes how it behaves for a NULL coming from a field, we should really consider. I guess we can always declare it a breaking change?

Comment (palesz) [reply x-pack/plugin/sql/qa/server/src/main/resources/functions.csv-spec]: I reverted to the alternative approach and made the `LOCATE(pattern,input,start=NULL{literal})` mimic the current behaviour of the `LOCATE(pattern,input,start=NULL{field})` in this PR. Our `LOCATE` also handles the negative `start` values differently from MySQL. `start=missing/NULL/negative value` all result the same as `start=1` (tries to find the pattern in the whole input) while MySQL considers `start=negative value/NULL` as invalid and returns with `0` (no match). Because `LOCATE` treats `NULL` as the `start` argument specially even in MySQL (not the same as any other argument being `NULL`) I will address these differences in a separate PR.

I am still adding the `breaking change` label to this PR, since the behaviour of the functions change anyways (`NULL` field values will start returning `NULL` values instead of the original input strings).

**Comment by mark-vieira** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+ * 
+ * <p>The test class based on the example function calls (and argument specifications) generates all the 
+ * permutations of the function arguments (4 options per argument: value/NULL as literal/field) and tests that the 
+ * function calls with the same argument values provide the same result regardless of the source (literal, field) 
+ * of the arguments.</p>
+ *  
+ * <p>To ignore any of the tests, add an .ignore() method call after the Fn ctors in the FUNCTION_CALLS_TO_TEST list below, like:
+ * <code> new Fn("ASCII", "foobar").ignore()</code></p>
+ */
+public class ConsistentFunctionArgHandlingIT extends JdbcIntegrationTestCase {
```
Comment (costin) [inline x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:49]: +1 for generating the test case to test various combinations however I find the current approach too convoluted.
Instead of indexing the data, filtering the functions and generating queries through an integration test one could use the `FunctionRegistry` directly and just focus on the string functions either by declaring their arguments as in this test or inspecting the returned definitions.

Specify the arguments or generate the loops based on the signature and pass them to the definition to get back the actual function instance to fold.

The resulting code should be smaller, without strings and faster due to be being a unit test.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java [hunk 3]
```diff
@@ -80,6 +81,11 @@ protected Pipe makePipe() {
         return NodeInfo.create(this, Locate::new, pattern, input, start);
     }
 
+    @Override
+    public Nullability nullable() {
+        return Nullability.FALSE;
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java:86]: If the function returns null if any argument is null, this should be `TRUE` not `FALSE`.

**Comment by costin** [comment]
Not only the functions need to be checked but also the underlying processors which are called for performing extractions or working on painless. This is needed to make sure that passing nulls to processors works the same as passing nulls to the functions.

While the processors are tested separately, it's worth looking into a test that checks both with the same set of arguments. At that point, manually declaring the functions and the processors plus their args might be easier than going for the discovery approach.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java (hunk 3 above)
Comment (palesz) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java:86]: `LOCATE` does not return `NULL` if `start=NULL`.

**Comment by matriv** [comment]
@palesz With the opportunity of this fix, please adjust the docs for those functions, where currently seem to not mention (apart from LOCATE) the edge cases with null args.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java (hunk 2 above)
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:49]: I agree that writing a unit test  would result in faster tests, however testing if:

1. `processors`
2. `Function.fold()` (optimizer folding rule that calls that if all args are foldable, calls the processor)
3. `optimizer rules` that rewrite the function call in the query (NULL folding for example that depends on `Function.nullable()`)
4. `QueryTranslator` output (function calls turn into painless scripts that call the processors within Query DSL queries)

all work together in a consistent way providing the same outcome regardless the argument sources being field/literal is easier through integration tests, because  the different argument source combinations will hit all of the above code paths using the same test code.

I agree though that in unit tests like [LocateProcessorTests](https://github.com/elastic/elasticsearch/blob/baa147cb7c5a39954c279d0876c49ad32f3d8312/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/LocateProcessorTests.java#L49) (or one similar to this in the PR) we could easily test if `fold()` and the `processor` leads to the same results, and likely it'll catch some of the bugs. But it will not cover `#3` and `#4`.

To cover `#3` we would have to make sure we call the `Optimizer` too, check if the function transformed by the rules will provide an output equivalent to `fold` and `processor` calls.
To cover `#4` we would have to execute the query generated by the `QueryTranslator` (or at least the generated Painless script), which I think is easier to do through integration tests and would require quite a bit of mocking in a unit test.

This looks too much of a whitebox testing for me. I'd rather execute the same code that the user's code would execute with less assumptions about our implementation. IMO it is easier to argue about the coverage of this integration test that uses the same approach to hit all the four cases above.

Discovery: `SqlFunctionRegistry` is not accessible in integration test (not impossible to fix), but even if it would be (or would want to use in a unit test): I don't see how can I easily figure out the type of the arguments from the `FunctionDefinition`. The constructor arguments of the functions are `Expression`s without hints about the actual accepted types and the `Builder`s also don't provide help. One way is to construct a function with generic/widely unacceptable arguments, call `resolveType()` and parse the error messages. It would help if the functions would be self-decribing (we could even generate  a function arguments meta table with `type`, `optional`, `accepted sources` per function that might be useful somewhere else), but it is outside of the scope of this PR.

**Comment by palesz** [comment]
> @palesz With the opportunity of this fix, please adjust the docs for those functions, where currently seem to not mention (apart from LOCATE) the edge cases with null args.

See https://github.com/elastic/elasticsearch/pull/68379/commits/7b8de11886004e42f4d154aa5a03f2bc74281245 of this PR. Clarified the edge cases on `CONCAT` and `LOCATE` (current behavior kept in this PR). The other functions don't have edge cases, any `NULL` args will return `NULL`, so left their doc untouched.

**Review by astefan — COMMENTED** [review]
Left one question mainly, the second one can be left unanswered as I'm not too convinced the alternative is better.
@costin has a point about the IT tests being maybe a bit too much for what they are achieving. There are ~500 queries being executed when the tests run. If the impact on the execution time is not too much, I'd keep them.
Otherwise LGTM.

**Comment by matriv** [comment]
> > @palesz With the opportunity of this fix, please adjust the docs for those functions, where currently seem to not mention (apart from LOCATE) the edge cases with null args.
> 
> See [7b8de11](https://github.com/elastic/elasticsearch/commit/7b8de11886004e42f4d154aa5a03f2bc74281245) of this PR. Clarified the edge cases on `CONCAT` and `LOCATE` (current behavior kept in this PR). The other functions don't have edge cases, any `NULL` args will return `NULL`, so left their doc untouched.

Thx! Though, myself, I'd prefer to explicitly state what's the behaviour with `NULL` in all cases, like we do for example in: https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-datetimeparse: 
```
Description: Returns a datetime by parsing the 1st argument using the format specified in the 2nd argument. The parsing format pattern used is the one from java.time.format.DateTimeFormatter. *If any of the two arguments is null or an empty string null is returned*.
```

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        private final String name;
+        private final List<Argument> arguments;
+        private List<String> aliases = new ArrayList<>();
+        private boolean ignored = false;
+        
+        private Fn(String name, Object... arguments) {
+            this.name = name;
+            this.arguments = new ArrayList<>();
+            for (Object a : arguments) {
+                this.arguments.add(Argument.class.isAssignableFrom(a.getClass()) ? (Argument) a : new Argument(a));
```
[the remark this answers is not in the document]
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:102]: Previously I used for one of the cases (`LOCATE`). Now this is unused and can be removed, but might come handy if we have a restriction that some of the arguments cannot come from fields.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java:86]: What about the other parameters?

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        Map<String, Object> testDoc = new LinkedHashMap<>();
+        testDoc.put("docId", testDocId);
+        int idx = 0;
+        for (Argument arg : fn.arguments) {
+            idx += 1;
+            testDoc.put(argPrefix + idx, arg.exampleValue);
+            // first set the same value, so the mapping is populated for the null columns
+            testDoc.put(nullArgPrefix + idx, arg.exampleValue);
+        }
+        index(indexName, functionName, body -> body.mapContents(testDoc));
```
[the remark this answers is not in the document]
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:225]: Was thinking along the same lines. It would definitely be more code.

**Comment by palesz** [comment]
> @costin has a point about the IT tests being maybe a bit too much for what they are achieving. There are ~500 queries being executed when the tests run.

@astefan See my previous comment https://github.com/elastic/elasticsearch/pull/68379#discussion_r569648252 about the reasoning behind integration tests. 

> If the impact on the execution time is not too much, I'd keep them.

The impact on execution time is about ~13 seconds. This test runs as part of the ci/2 that takes ~60 minutes.

**Comment by palesz** [comment]
> Thx! Though, myself, I'd prefer to explicitly state what's the behaviour with `NULL` in all cases

As discussed, will handle in https://github.com/elastic/elasticsearch/issues/68531

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java [hunk 6]
```diff
@@ -48,21 +48,15 @@ public Object process(Object input) {
...
             return null;
         }
         if ((input instanceof String || input instanceof Character) == false) {
             throw new SqlIllegalArgumentException("A string/char is required; received [{}]", input);
         }
-        if (replacement == null) {
-            return input;
-        }
-        if ((replacement instanceof String || replacement instanceof Character) == false) {
+        if (!(replacement instanceof String || replacement instanceof Character)) {
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java:57]: Nit: was the change intentional (here and one more spot in the patch)?
(Myself I dislike `== false`, but since this the direction [the train is going](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines), I'm comfier with a consistent style.)

**Review by bpintea — APPROVED** [review]
LGTM.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java (hunk 6 above)
Comment (palesz) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java:57]: Thanks for pointing out. No, it wasn't, I think just IDEA tried to be helpful (will revert this one).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java (hunk 3 above)
Comment (palesz) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java:86]: If any of the `pattern` or `input` is `NULL, the result is `NULL`. Only `start` is special.

Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java:86]: Then the function is nullable as oppose to never nullable. Depending on the params (outside start) it can TRUE or UKNOWN - see `Nullability.and`.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:102]: @palesz I would suggest removing it and use it when needed. Someone looking at the code gets confused about a functionality that's not actually there.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java [hunk 7]
```diff
@@ -193,18 +194,17 @@ public void test() throws Exception {
...
                 final String query = "SELECT " + functionCall + " FROM " + indexName + " WHERE docId = '" + testDocId + "'";
                 ResultSet retVal = esJdbc().createStatement().executeQuery(query);
-                
+
                 assertTrue(retVal.next());
                 results.add(tuple(functionName + "(" + join(", ", functionCallArgsForAssert) + ")", retVal.getObject(1)));
                 // only a single row should be returned
                 assertFalse(retVal.next());
 
-                Stream<Object> returnValueStream = results.stream().map(Tuple::v2);
```
Comment (astefan) [inline x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java]: Why did you remove this and now go over the stream components twice?

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        new Fn("SUBSTRING", "foobar", 1, 2),
+        new Fn("TRIM", "  foobar   "),
+        new Fn("UCASE", "foobar")
+    );
+
+    private static final List<String> NON_TESTED_FUNCTIONS;
+    static {
+        try {
+            Class<?> c = ConsistentFunctionArgHandlingIT.class;
+            NON_TESTED_FUNCTIONS = Files.readAllLines(Path.of(c.getResource(c.getSimpleName() + "-non-tested-functions.txt").toURI()));
```
Comment (matriv) [inline x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:81]: It would be great if it's possible get all the functions from FunctionRegistry:
Something like: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java#L2199

Can also be done at a later point and not as part of this PR.

**Review by matriv — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java (hunk 7 above)
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java]: Sadly streams can only be iterated once, during the second iteration it gave `stream is already closed` exception.

Comment (astefan) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java]: Why did the tests pass then?

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java (hunk 8 above)
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:81]: `SqlFunctionRegistry` is not available in integration tests (can add it as a dependency, but I don't think we really need it). I get the list of the functions from `SHOW FUNCTIONS` [see here](https://github.com/palesz/elasticsearch/blob/4e20cbf0f8223c5be773c25b58e97de2b03d8fa3/x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java#L235), that achieves pretty much the same that `SqlFunctionRegistry` would (for more see the last paragraph [here](https://github.com/elastic/elasticsearch/pull/68379#discussion_r569648252)). The purpose of the `NON_TESTED_FUNCTIONS` list is to identify functions added after writing this test class which likely we should think about adding to the tested function list / non tested function list (one-liner). Newly added = `SHOW FUNCTIONS \ FUNCTIONS_CALLS \ NON_TESTED_FUNCTIONS` should be an empty set. The `NON_TESTED_FUNCTIONS` list should shrink later once the datetime, ... functions are also added to the tested function list (separate PR).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java (hunk 3 above)
Comment (palesz) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java:86]: You are right, changed.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java:102]: Removed.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java (hunk 6 above)
Comment (palesz) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java]: Reverted.

File: x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java (hunk 7 above)
Comment (palesz) [reply x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java]: Because the second iteration only happens if the test finds an issue (to create the detailed message for error).

**Review by astefan — APPROVED** [review]
LGTM
